File: Assets/Scripts/messages/Score.js
#pragma strict

/**
 * @author: cmdr2 <secondary.cmdr2@gmail.com>
 */

private static var instance : Score;

/* scratchpad */
private var scoreText : TextMesh;
private var score : int = 0;
private var levelScore : int = 0;
private var lives : int = 0;


@script RequireComponent(TextMesh)
function Start () {
	scoreText = GetComponent(TextMesh);
	Render();
}


public static function Instance() : Score {
	if (instance == null) {
		instance = UnityEngine.Object.FindObjectOfType(Score);
	}
	return instance;
}


public function AddScore(amount : int) {
	SetScore(score + amount);
}


public function AddLives(amount : int) {
	SetLives(lives + amount);
}


public function SetScore(amount : int) {
	score = (amount < 0 ? 0 : amount);
	Render();
}


public function SetLevelScore(amount : int) {
	levelScore = (amount < 0 ? 0 : amount);
	Render();
}


public function SetLives(amount : int) {
	lives = (amount < 0 ? 0 : amount);
	Render();
}


private function Render() {
	if (scoreText) {
		if (score + levelScore > 0) {
			scoreText.text = 'Score: ' + (score + levelScore);
		} else {
			scoreText.text = '';
		}
	} else {
		Metrics.LogException("Score: Render(): scoreText is null", true);
	}
}

File: Assets/Scripts/messages/Score.cs
using UnityEngine;
using System.Collections;

[System.Serializable]
/**
 * @author: cmdr2 <secondary.cmdr2@gmail.com>
 */
/* scratchpad */
[UnityEngine.RequireComponent(typeof(TextMesh))]
public partial class Score : MonoBehaviour
{
    private static Score instance;
    private TextMesh scoreText;
    private int score;
    private int levelScore;
    private int lives;
    public virtual void Start()
    {
        this.scoreText = (TextMesh) this.GetComponent(typeof(TextMesh));
        this.Render();
    }

    public static Score Instance()
    {
        if (Score.instance == null)
        {
            Score.instance = (Score) UnityEngine.Object.FindObjectOfType(typeof(Score));
        }
        return Score.instance;
    }

    public virtual void AddScore(int amount)
    {
        this.SetScore(this.score + amount);
    }

    public virtual void AddLives(int amount)
    {
        this.SetLives(this.lives + amount);
    }

    public virtual void SetScore(int amount)
    {
        this.score = amount < 0 ? 0 : amount;
        this.Render();
    }

    public virtual void SetLevelScore(int amount)
    {
        this.levelScore = amount < 0 ? 0 : amount;
        this.Render();
    }

    public virtual void SetLives(int amount)
    {
        this.lives = amount < 0 ? 0 : amount;
        this.Render();
    }

    private void Render()
    {
        if (this.scoreText)
        {
            if ((this.score + this.levelScore) > 0)
            {
                this.scoreText.text = "Score: " + (this.score + this.levelScore);
            }
            else
            {
                this.scoreText.text = "";
            }
        }
        else
        {
            Metrics.LogException("Score: Render(): scoreText is null", true);
        }
    }

}